### PR TITLE
Add workday signature suite, plus examples

### PR DESCRIPTION
**Pull request by thereisnogabe**
My hope is to temporarily host the code sample and document outlining the signature suite until we are able to stand up our own hosting (it's an unfortunately slow process).

### CODE REVIEW

**Comment by msporny** [comment]
> My hope is to temporarily host the code sample and document outlining the signature suite until we are able to stand up our own hosting (it's an unfortunately slow process).

Hmm, this is going to be a problem. Can't you host the sample code and document for the signature suite somewhere else on Github? Even on a personal account? The LD Cryptosuite Registry was never intended to host sample code or specs, just references to specs (which could then point to sample implementations).

We can't merge due to these reasons, we should discuss on an upcoming W3C CCG call if you'd like. /cc @kimdhamilton @ChristopherA @jandrieu 

I did a quick review of the code and spec since it is here and have a number of security-related concerns:

* Your canonicalization algorithm seems to not take into account floating point representation formats, integer sizes, and other items that could cause signatures to falsely validate on devices with restricted integer/float representation sizes. More specifically:
  * You don't seem to specify how duplicate property names should be handled.
  * You don't seem to specify how non-Unicode strings might be handled.
  * You don't seem to specify how double precision numbers should be handled
* It looks like the canonicalization algorithm only works on JSON, which is dangerous in the event that some people might think it will also work on JSON-LD w/o issue. This could lead to different canonical forms for the same information model, and signature formats that are not syntax agnostic.
* I didn't see a mention of [JCS](https://tools.ietf.org/id/draft-rundgren-json-canonicalization-scheme-10.html), which you should probably use instead of the bespoke canonicalization mechanism shown here.

Just flagging those as preliminary concerns based on a 5 minute read of this PR.

Please let us know how you'd like to proceed.

**Comment by thereisnogabe** [comment]
Quickly:
1. We only intend to support unicode strings with this algorithm (at this time).
2. We only intend to support JSON with this algorithm (at this time).
3. We are specifying this because we do not wish to use JCS or any other canonicalization algorithm (at this time).
4. For floating point & duplicate, we defer to JSON handling (at this time).

Hosting, if it's a big sticking point, I will bring it back to the team to find another solution.

**Comment by msporny** [comment]
> 1. We only intend to support unicode strings with this algorithm (at this time).

Yes, but remember that there are JSON implementations out there that support other types of strings or are limited. This creates a potential security vulnerability for your signature algorithm.

> 2. We only intend to support JSON with this algorithm (at this time).

That's fine, just making sure that you know the downsides of that approach.

> 3. We are specifying this because we do not wish to use JCS or any other canonicalization algorithm (at this time).

Hmm, this one is going to be frowned upon. You are defining a new, less capable canonicalization algorithm that is prone to security vulnerabilities.

To be perfectly clear, you can do that and still get into the registry. However, if you intend this to get standardized in any way, you're in for a world of pain when you try to convince the security review team why your mechanism is a better choice than JCS.

> 4. For floating point & duplicate, we defer to JSON handling (at this time).

JSON doesn't specify how to handle those issues... the spec is silent... that's the problem.

> Hosting, if it's a big sticking point, I will bring it back to the team to find another solution.

Yes, we don't host things in the LD Cryptosuite Registry. All that's required is that you create a new CCG repo and host it there. Here's an example:

https://github.com/w3c-ccg/lds-ed25519-2018

I expect that the CCG would allow you to do that and that shouldn't create issues Workday.

Again -- my review of your mechanism has nothing to do w/ integration into the registry - that'll happen after some of the more administrative refactoring I mentioned above. I was just reviewing it as a courtesy and pointing out the areas that you may want to consider more deeply as you develop this mechanism.

**Comment by thereisnogabe** [comment]
Thank you Manu. I appreciate you taking the time to give us feedback, it's great to have your expertise weighing in. We need to spend some more time examining this. I will work with my team and post a reply here next week. Will move the code to a new repo.

**Comment by thereisnogabe** [comment]
Hi @msporny if we modify the proposal to use JCS (and move to our own hosting) would we be on track for acceptance?

**Comment by OR13** [comment]
An example using JCS in javascript... https://github.com/w3c-ccg/vc-json-schemas/blob/master/hypothetical-use-cases/non-jsonld/src/__tests__/no-jsonld.spec.js#L30

This work should be hosted in a seperate repo, and linked... see the examples here:

https://github.com/w3c-ccg/lds-jws2020

https://gpg.jsld.org/

**Review by OR13 — CHANGES_REQUESTED** [review]
This is a registry, not a code repo.

**Comment by OR13** [comment]
@glcohen ping me in slack if you have questions about this, I'm pretty confident we can get it added if you base your approach off the examples I linked above (already in the registry).

File: doc/workday/signature-suite.md
Comment (OR13) [inline doc/workday/signature-suite.md:34]: You might want to rephrase this as "This specification is intended to not rely on JSON-LD canonicalization"... JSON-LD is JSON.

Comment (thereisnogabe) [reply doc/workday/signature-suite.md:34]: Thanks, will rephrase

File: doc/workday/signature-suite.md
Comment (OR13) [inline doc/workday/signature-suite.md:63]: only Ed25519? Consider adding support for JOSE... https://github.com/w3c-ccg/lds-jws2020

Ed25519 is great, but there are other key types and signature formats... and you will need to create a new suite for each if you make this only Ed25519... You will need to register each of them again, and the PRs to get them added to DID Core / W3C might take months....

Comment (OR13) [reply doc/workday/signature-suite.md:63]: Also you probably need to specify "Detached" or not detached "EdDSA"... I expect "Detached" given thats what every othter LD suite has used.

File: index.html
Comment (OR13) [inline index.html:595]: this does not look like a detached jws.... this is fine if thats your intention.

File: index.html
Comment (OR13) [inline index.html:562]: Possibility to allow other key types in the future here... also depending on how you are representing key types, you can add support for existing suites key types, by being explicit.

here are some suites which have key types that would be compatible:

```
{
      "id": "did:example:123#_Qq0UL2Fq651Q0Fjd6TvnYE-faHiOpRlPVQcY_-tA4A",
      "type": "JwsVerificationKey2020",
      "controller": "did:example:123",
      "publicKeyJwk": {
        "crv": "Ed25519",
        "x": "VCpo2LMLhn6iWku8MKvSLg2ZAoC-nlOyPVQaO3FxVeQ",
        "kty": "OKP",
        "kid": "_Qq0UL2Fq651Q0Fjd6TvnYE-faHiOpRlPVQcY_-tA4A"
      }
    },
```

```
 {
      "id": "did:key:z6MkuVK4wPm1fFaovcHxRTP6S8chkQuNCdvzuVZHfzVJWGU7#z6MkuVK4wPm1fFaovcHxRTP6S8chkQuNCdvzuVZHfzVJWGU7",
      "type": "Ed25519VerificationKey2018",
      "controller": "did:key:z6MkuVK4wPm1fFaovcHxRTP6S8chkQuNCdvzuVZHfzVJWGU7",
      "publicKeyBase58": "G342M9WaKi6Lp7TFjtRFb34hvqdWnkgeDUeMqiXHb3gj"
    }
```

**Comment by thereisnogabe** [comment]
Closing, will re-open when changes are addressed.

**Pull request closed without merging**
